src/Cardano/Wallet/CoinSelection/Fee.hs

```diff
@@ -0,0 +1,351 @@
+            :: ([Coin] -> [Coin])
+            -> Fee
+            -> [Coin]
+            -> ([Coin], Fee)
+        reduceChangeOutputs f totalFee chgs =
+            case (filter (/= Coin 0) . f) chgs of
+                [] ->
+                    (f chgs, totalFee)
+                xs ->
+                    bimap removeDust (Fee . Coin . sum . map (getCoin . getFee))
+                    $ L.unzip
+                    $ map reduceSingleChange
+                    $ divvyFee totalFee xs
+
+        reduceSingleChange
+            :: (Fee, Coin)
+            -> (Coin, Fee)
+        reduceSingleChange (fee, chng)
+            | chng >= getFee fee =
+                  (Coin $ (getCoin chng) - ((getCoin . getFee) fee), (Fee . Coin) 0 )
+            | otherwise =
+                  (Coin 0, Fee . Coin $ ((getCoin . getFee) fee) - (getCoin chng) )
+
+
+        go
+            :: forall m. MonadRandom m
+            => CoinSelection
+            -> ExceptT FeeError m CoinSelection
+        go coinSel@(CoinSelection inps outs chgs) = do
+            -- 1/
+            -- We compute fees using all inputs, outputs and changes since
+            -- all of them have an influence on the fee calculation.
+            let upperBound = feeUpperBound opt coinSel
+
+            -- 2/ Substract fee from all change outputs, proportionally to their value.
+            let (chgs', remainingFee) =
+                    reduceChangeOutputs removeDust upperBound chgs
+
+            -- 3.1/
+            -- Should the change cover the fee, we're (almost) good. By removing
+            -- change outputs, we make them smaller and may reduce the size of the
+            -- transaction, and the fee. Thus, we end up paying slightly more than
+            -- the upper bound. We could do some binary search and try to
+            -- re-distribute excess across changes until fee becomes bigger.
+            if (getFee remainingFee) == Coin 0 then
+                pure $ CoinSelection inps outs chgs'
+            else do
+            -- 3.2/
+            -- Otherwise, we need an extra entries from the available utxo to
+            -- cover what's left. Note that this entry may increase our change
+            -- because we may not consume it entirely. So we will just split
+            -- the extra change across all changes possibly increasing the
+            -- number of change outputs (if there was none, or if increasing
+            -- a change value causes an overflow).
+            --
+            -- Because selecting a new input increases the fee, we need to
+            -- re-run the algorithm with this new elements and using the initial
+            -- change plus the extra change brought up by this entry and see if
+            -- we can now correctly cover fee.
+                remFee <- lift $ runMaybeT $ coverRemainingFee remainingFee utxo
+                case remFee of
+                    Nothing ->
+                        throwE CannotCoverFee
```

Comment:
Maybe we can have the error a bit more explicit here :thinking: ? Saying that we cannot cover for fee because there's not enough inputs? And actually communicate about the missing amount?